workitem/workitem_repository.go

```diff
@@ -85,6 +103,110 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
 	return nil
 }
 
+// Reorder places the to-be-reordered workitem(s) above the input workitem.
+// The order of workitems are spaced by a factor of 1000.
+// The new order of workitem := (order of previousitem + order of nextitem)/2
+// Version must be the same as the one int the stored version
+func (r *GormWorkItemRepository) Reorder(ctx context.Context, before string, wi app.WorkItem) (*app.WorkItem, error) {
+	var order float64
+	res := WorkItem{}
+	beforeItem := WorkItem{}
+	afterItem := WorkItem{}
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil || id == 0 {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+
+	tx := r.db.First(&res, id)
+	if tx.RecordNotFound() {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if res.Version != wi.Version {
+		return nil, errors.NewVersionConflictError("version conflict")
+	}
+
+	wiType, err := r.wir.LoadTypeFromDB(wi.Type)
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+
+	if before != "" {
+		beforeId, err := strconv.ParseUint(before, 10, 64)
+		if err != nil || beforeId == 0 {
+			return nil, errors.NewNotFoundError("work item", before)
+		}
+		tx = r.db.First(&beforeItem, beforeId)
+		if tx.RecordNotFound() {
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		if tx.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		beforeOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", beforeItem.Fields[SystemOrder]), 64)
+		if err != nil {
+			return nil, errors.NewBadParameterError("data.attributes.order", res.Fields[SystemOrder])
+		}
+		tx2 := r.db.Where("fields -> 'order' < ?", beforeItem.Fields[SystemOrder]).Order("fields->'order' desc", true).Last(&afterItem)
```

Comment:
The use of fields -> '' won't hit the index. It will result in a full table scan as is. Some other solution is needed. The easiest is probably to move order to a column and not be part of the fields blob. I'll do some investigation on this.